components/script/dom/worklet.rs

```diff
@@ -0,0 +1,621 @@
+    }
+
+    /// Get the worklet global scope for a given worklet.
+    /// Creates the worklet global scope if it doesn't exist.
+    fn get_worklet_global_scope(&mut self,
+                                pipeline_id: PipelineId,
+                                worklet_id: WorkletId,
+                                global_type: WorkletGlobalScopeType,
+                                base_url: ServoUrl)
+                                -> Root<WorkletGlobalScope>
+    {
+        match self.global_scopes.entry(worklet_id) {
+            hash_map::Entry::Occupied(entry) => Root::from_ref(entry.get()),
+            hash_map::Entry::Vacant(entry) => {
+                let result = global_type.new(&self.runtime, pipeline_id, base_url, &self.global_init);
+                entry.insert(JS::from_ref(&*result));
+                result
+            },
+        }
+    }
+
+    /// Fetch and invoke a worklet script.
+    ///g https://drafts.css-houdini.org/worklets/#fetch-and-invoke-a-worklet-script
+    fn fetch_and_invoke_a_worklet_script(&self,
+                                         global_scope: &WorkletGlobalScope,
+                                         script_url: ServoUrl,
+                                         credentials: RequestCredentials,
+                                         pending_tasks_struct: PendingTasksStruct,
+                                         promise: TrustedPromise)
+    {
+        debug!("Fetching from {}.", script_url);
+        // Step 1.
+        // TODO: Settings object?
+
+        // Step 2.
+        // TODO: Fetch a module graph, not just a single script.
+        // TODO: Fetch the script asynchronously?
+        // TODO: Caching.
+        let resource_fetcher = self.global_init.resource_threads.sender();
+        let request = RequestInit {
+            url: script_url,
+            type_: RequestType::Script,
+            destination: Destination::Script,
+            credentials_mode: credentials.into(),
+            .. RequestInit::default()
+        };
+        let script = load_whole_resource(request, &resource_fetcher).ok()
+            .and_then(|(_, bytes)| String::from_utf8(bytes).ok());
+
+        // Step 4.
+        // NOTE: the spec parses and executes the script in separate steps,
+        // but our JS API doesn't separate these, so we do the steps out of order.
+        let ok = script.map(|script| global_scope.evaluate_js(&*script)).unwrap_or(false);
+
+        if !ok {
+            // Step 3.
+            debug!("Failed to load script.");
+            let old_counter = pending_tasks_struct.set_counter_to(-1);
+            if old_counter > 0 {
+                self.run_in_script_thread(promise.reject_runnable(Error::Abort));
+            }
+        } else {
+            // Step 5.
+            debug!("Finished adding script.");
+            let old_counter = pending_tasks_struct.decrement_counter_by(1);
+            if old_counter == 1 {
+                // TODO: trigger a reflow?
```

Comment:
At this point, we have loaded a worklet, but not actually run it (e.g. to get a paint worklet to do its painting).